compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -257,76 +257,34 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     ) -> InterpResult<'tcx> {
-        // A<Struct> -> A<Trait> conversion
-        let (src_pointee_ty, dest_pointee_ty) =
-            self.tcx.struct_lockstep_tails_erasing_lifetimes(source_ty, cast_ty, self.param_env);
-
-        match (&src_pointee_ty.kind(), &dest_pointee_ty.kind()) {
-            (&ty::Array(_, length), &ty::Slice(_)) => {
-                let ptr = self.read_immediate(src)?.to_scalar()?;
-                // u64 cast is from usize to u64, which is always good
-                let val =
-                    Immediate::new_slice(ptr, length.eval_usize(*self.tcx, self.param_env), self);
-                self.write_immediate(val, dest)
-            }
-            (&ty::Dynamic(ref data_a, ..), &ty::Dynamic(ref data_b, ..)) => {
-                let val = self.read_immediate(src)?;
-                if data_a.principal_def_id() == data_b.principal_def_id() {
-                    return self.write_immediate(*val, dest);
-                }
-                // trait upcasting coercion
-                let vptr_entry_idx = self.tcx.vtable_trait_upcasting_coercion_new_vptr_slot((
-                    src_pointee_ty,
-                    dest_pointee_ty,
-                ));
+        let (data, old_meta) = match *self.read_immediate(src)? {
+            Immediate::Scalar(data) => (data.check_init()?, ScalarMaybeUninit::Uninit),
+            Immediate::ScalarPair(data, meta) => (data.check_init()?, meta),
+        };
 
-                if let Some(entry_idx) = vptr_entry_idx {
-                    let entry_idx = u64::try_from(entry_idx).unwrap();
-                    let (old_data, old_vptr) = val.to_scalar_pair()?;
-                    let old_vptr = self.scalar_to_ptr(old_vptr)?;
-                    let new_vptr = self
-                        .read_new_vtable_after_trait_upcasting_from_vtable(old_vptr, entry_idx)?;
-                    self.write_immediate(Immediate::new_dyn_trait(old_data, new_vptr, self), dest)
-                } else {
-                    self.write_immediate(*val, dest)
-                }
-            }
-            (_, &ty::Dynamic(ref data, _)) => {
-                // Initial cast from sized to dyn trait
-                let vtable = self.get_vtable(src_pointee_ty, data.principal())?;
-                let ptr = self.read_immediate(src)?.to_scalar()?;
-                let val = Immediate::new_dyn_trait(ptr, vtable, &*self.tcx);
-                self.write_immediate(val, dest)
-            }
-
-            _ => {
-                span_bug!(self.cur_span(), "invalid unsizing {:?} -> {:?}", src.layout.ty, cast_ty)
-            }
-        }
+        let new_meta = self.unsize_just_metadata(old_meta, source_ty, cast_ty)?;
+        self.write_immediate(Immediate::ScalarPair(data.into(), new_meta.into()), dest)
     }
 
     fn unsize_just_metadata(
         &mut self,
-        src: &OpTy<'tcx, M::PointerTag>,
-        dest: &PlaceTy<'tcx, M::PointerTag>,
+        src: ScalarMaybeUninit<M::PointerTag>,
```

Comment:
This is the source *metadata*, right? `src_meta` might be a better name then.

Reply:
👍